CRM/ManualDirectDebit/Hook/PostProcess/Membership/DirectDebitMandate.php

```diff
@@ -67,6 +72,44 @@ private function setMandateValues() {
     }
   }
 
+  /**
+   * Checks if 'directDebitMandate_authorisation_file' was added in mandate
+   *
+   * @param $submitFiles
+   *
+   * @return bool
+   */
+  private function isFileAttached($submitFiles) {
+    return !empty($submitFiles['tmp_name']);
+  }
+
+  /**
+   * Save file and assign it Id to mandate
+   *
+   * @param $file
+   */
+  private function setMandateFile($file) {
+    CRM_Core_BAO_File::filePostProcess(
+      $file['tmp_name'],
+      NULL,
+      CRM_ManualDirectDebit_Common_MandateStorageManager::DIRECT_DEBIT_TABLE_NAME,
+      $this->currentContactId,
+      NULL,
+      TRUE,
+      NULL,
+      'uploadedMandateFile',
+      $file['type']
+    );
+
+    $sqlSelectDebitMandateID = "SELECT MAX(id) as id FROM `civicrm_file`";
+
+    $queryResult = CRM_Core_DAO::executeQuery($sqlSelectDebitMandateID);
+    $queryResult->fetch();
+    if (isset($queryResult->id) && !empty($queryResult->id)){
+      $this->mandateValues['authorisation_file'] = $queryResult->id;
+    }
```

Comment:
Please embed this whole function in a transaction, just in case, to protect against race conditions from file post processing and MAX(id) obtaining.

Eg.
 ```php
private function setMandateFile($file) {
  $transaction = new CRM_Core_Transaction();

  try {
    CRM_Core_BAO_File::filePostProcess(...);

    $sqlSelectDebitMandateID = "SELECT MAX(id) as id FROM `civicrm_file`";
    $queryResult = CRM_Core_DAO::executeQuery($sqlSelectDebitMandateID);
    $queryResult->fetch();

    if (isset($queryResult->id) && !empty($queryResult->id)){
      $this->mandateValues['authorisation_file'] = $queryResult->id;
    }
  } catch (Exception $exception) {
    $transaction->rollback();
    throw $exception;
  }
}
```